internal/twirptest/proto/proto_test.go

```diff
@@ -13,8 +13,129 @@
 
 package proto
 
-import "testing"
+import (
+	context "context"
+	fmt "fmt"
+	ioutil "io/ioutil"
+	http "net/http"
+	"net/http/httptest"
+	strings "strings"
+	"testing"
+	"time"
+
+	twirp "github.com/twitchtv/twirp"
+)
 
 func TestCompilation(t *testing.T) {
 	// Test passes if this package compiles
 }
+
+type errReader int
+
+func (errReader) Read(p []byte) (n int, err error) {
+	return 0, fmt.Errorf("test error")
+}
+
+func TestContextCancelError(t *testing.T) {
+	// Test passes if this package compiles
+	var s Svc
+	server := NewSvcServer(s, nil)
+
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+
+	// Make a request, that will endpoint method
+	req, _ := http.NewRequest(http.MethodPost, "http://testing:8080/twirp/twirp.internal.twirptest.proto.Svc/Send", errReader(0))
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
+	// Associate the cancellable context we just created to the request
+	req = req.WithContext(ctx)
+	// cancel context
+	cancel()
+	w := httptest.NewRecorder()
+	server.ServeHTTP(w, req)
+
+	expectedErrCode := twirp.ServerHTTPStatusFromErrorCode(twirp.Canceled)
+	if w.Code != expectedErrCode {
+		t.Errorf("twirp ErrorCode expected to be %q, but found %q", expectedErrCode, w.Code)
+	}
+
+	respBytes, err := ioutil.ReadAll(w.Body)
+	if err != nil {
+		t.Fatalf("Could not even read bytes from response: %q", err.Error())
+	}
+
+	expectedErrMessage := "context canceled"
+	if !strings.Contains(string(respBytes), expectedErrMessage) {
+		t.Errorf("twirp client err has unexpected message %q, want %q", string(respBytes), expectedErrMessage)
+	}
+}
+
+func TestDeadlineExceededError(t *testing.T) {
+	// Test passes if this package compiles
+	var s Svc
+	server := NewSvcServer(s, nil)
+
+	ctx := context.Background()
+	ctx, _ = context.WithTimeout(ctx, 1*time.Millisecond)
+
+	// Make a request, that will endpoint method
+	req, _ := http.NewRequest(http.MethodPost, "http://testing:8080/twirp/twirp.internal.twirptest.proto.Svc/Send", errReader(0))
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
+	// Associate the cancellable context we just created to the request
+	req = req.WithContext(ctx)
+	// cancel context
+	time.Sleep(2 * time.Millisecond)
```

Comment:
Instead of using time.Sleep, you should check the timeout channel like this:
```go
select {
case <-ctx.Done():
    // ctx.Err() will have context.DeadlineExceeded
case <-time.After(1 * time.Second):
    t.Fatalf("Waited too long for DeadlineExceeded ")
}
```